### PR TITLE
Implement parallel worker in machine controller.

### LINKED ISSUE #101: Controller reconciliation loop is serialized
_From @medinatiger on April 3, 2018 17:21_

Machine controller reconciliation is serialized. We need to find a way to make operation in parallel.

_Copied from original issue: kubernetes/kube-deploy#663_

### PR DESCRIPTION
**Pull request by medinatiger**
**What this PR does / why we need it**:
This PR add multiple workers for machine controller loop which make machine reconciliation loops in parallel for different machine resource.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #101 

**Special notes for your reviewer**:
This is a migration for PR: https://github.com/kubernetes/kube-deploy/pull/681. Thanks for Martin's explanation for how client-go queue works, it makes the solution much simpler! The ultimate goal is still to incorpate it into apiserver-builder so that other controller can benefit from it too.

**Release note**:

```release-note

```

@kubernetes/kube-deploy-reviewers

### CODE REVIEW

File: pkg/controller/config/configuration.go [hunk 1]
```diff
@@ -21,15 +21,18 @@ import (
 )
 
 type Configuration struct {
-	Kubeconfig string
-	InCluster  bool
+	Kubeconfig  string
+	InCluster   bool
+	ThreadCount int
```
Comment (krousey) [inline pkg/controller/config/configuration.go]: nit: call this WorkerCount

File: pkg/controller/machine/queue.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+// StartWorkerQueue schedules a routine to continuously process Queue messages
+// until shutdown is closed
+func (c *MachineController) StartWorkerQueue(q *controller.QueueWorker, shutdown <-chan struct{}) {
+	glog.Infof("Start %s Queue", q.Name)
+
+	for i := 0; i < config.ControllerConfig.ThreadCount; i++ {
+		// Every second, process all messages in the Queue until it is time to shutdown
+		go wait.Until(func() { q.ProcessAllMessages() }, time.Second, shutdown)
```
Comment (krousey) [inline pkg/controller/machine/queue.go]: This is significantly simpler than what you had at first. Could you briefly explain why what you had before is not necessary and this is sufficient? Does the queue make sure a machine can't be queued more than once? It looks like maybe, but I'm not sure.

Also, do you need the anonymous function as the first argument to `wait.Until`? Or can you pass `q.ProcessAllMessages` directly?

Comment (medinatiger) [reply pkg/controller/machine/queue.go]: The queue package in client-go actually provided native support by allowing only one reconciliation active per resource. Checking out the Add() routine (https://github.com/kubernetes/client-go/blob/master/util/workqueue/queue.go), we don't really enqueue a new task if it's already in-progress. Instead, we marked the dirty bit. In Done() routine, we will requeue the task if it's marked as dirty. It's quite a clever and straightforward design.

File: pkg/controller/config/configuration.go (hunk 1 above)
Comment (medinatiger) [reply pkg/controller/config/configuration.go]: Done

File: pkg/controller/machine/queue.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+// until shutdown is closed
+func (c *MachineController) StartWorkerQueue(q *controller.QueueWorker, shutdown <-chan struct{}) {
+	glog.Infof("Start %s Queue", q.Name)
+
+	for i := 0; i < config.ControllerConfig.WorkerCount; i++ {
+		// Every second, process all messages in the Queue until it is time to shutdown
+		go wait.Until(q.ProcessAllMessages, time.Second, shutdown)
+	}
+
+	go func() {
```
Comment (mvladev) [inline pkg/controller/machine/queue.go]: I think you can make it simpler - you can run the entire `c.StartWorkerQueue(q, stopCh)` as goroutine - `go c.StartWorkerQueue(q, stopCh)`, defer the queue shutdown at the beginning of the function and block on the `shutdown` channel

File: pkg/controller/machine/queue.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+	"github.com/kubernetes-incubator/apiserver-builder/pkg/controller"
+	"sigs.k8s.io/cluster-api/pkg/controller/config"
+)
+
+func (c *MachineController) RunAsync(stopCh <-chan struct{}) {
+	for _, q := range c.Informers.WorkerQueues {
+		c.StartWorkerQueue(q, stopCh)
+	}
+	controller.GetDefaults(c.controller).Run(stopCh)
```
Comment (mvladev) [inline pkg/controller/machine/queue.go]: This does nothing. It calls https://github.com/kubernetes-sigs/cluster-api/blob/df991ef535792b2bbec5242d63316cfd153ab5e6/vendor/github.com/kubernetes-incubator/apiserver-builder/pkg/controller/controller.go#L131-L137 
which checks for 
https://github.com/kubernetes-sigs/cluster-api/blob/df991ef535792b2bbec5242d63316cfd153ab5e6/vendor/github.com/kubernetes-incubator/apiserver-builder/pkg/controller/controller.go#L139-L141
`Run(stopCh <-chan struct{})` function on our `c.controller` struct. We don't have such function, so it fallbacks to dummy default function.

Comment (medinatiger) [reply pkg/controller/machine/queue.go]: I understood that. My original intention was to keep consistency with api-server builder. Remove that.

File: pkg/controller/machine/queue.go (hunk 3 above)
Comment (medinatiger) [reply pkg/controller/machine/queue.go]: Done

File: pkg/controller/machine/queue.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+func (c *MachineController) RunAsync(stopCh <-chan struct{}) {
+	for _, q := range c.Informers.WorkerQueues {
+		go c.StartWorkerQueue(q, stopCh)
+	}
+}
+
+// StartWorkerQueue schedules a routine to continuously process Queue messages
+// until shutdown is closed
+func (c *MachineController) StartWorkerQueue(q *controller.QueueWorker, shutdown <-chan struct{}) {
+	glog.Infof("Start %s Queue", q.Name)
```
Comment (mvladev) [inline pkg/controller/machine/queue.go]: @medinatiger I think that this is a more simple solution

```go
glog.Infof("Start %s Queue", q.Name)
defer q.Queue.ShutDown()

for i := 0; i < config.ControllerConfig.WorkerCount; i++ {
  // Every second, process all messages in the Queue until it is time to shutdown
  go wait.Until(q.ProcessAllMessages, time.Second, shutdown)
}

<-shutdown

// Stop accepting messages into the Queue
glog.V(1).Infof("Shutting down %s Queue\n", q.Name)
```

Comment (medinatiger) [reply pkg/controller/machine/queue.go]: Done

**Comment by krousey** [comment]
Lgtm from what I could see on my phone.

**Comment by medinatiger** [comment]
Added a unit test.

File: pkg/controller/machine/controller_test.go [hunk 6]
```diff
@@ -94,3 +90,27 @@ func machineControllerReconcile(t *testing.T, cs *clientset.Clientset, controlle
...
+		instance := v1alpha1.Machine{}
+		instance.Name = "instance" + strconv.Itoa(i)
+		if _, err := client.Create(&instance); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	time.Sleep(2 * time.Second)
+
+	// Sometime the reconcilation will trigger the idempotent action which will not call into
```
Comment (jessicaochen) [inline pkg/controller/machine/controller_test.go]: Depending on timing in tests is the root of a lot of flaky tests. Can you have this test not depend on timing? For example, you can block one of the instances from completing reconciliation (eg. with channels) and see the others still get processed.

File: pkg/controller/machine/controller_test.go [hunk 7]
```diff
@@ -94,3 +90,27 @@ func machineControllerReconcile(t *testing.T, cs *clientset.Clientset, controlle
...
+	// actuator. 3 (larger than 2) is good enough to verify parallel execution.
+	if ta.CreateCallCount <= 3 {
+		t.Fatalf("Expect 5 completion, get %d", ta.CreateCallCount)
```
Comment (jessicaochen) [inline pkg/controller/machine/controller_test.go]: The number expected in the comment (5) does not match the number in the code (3). Can these be made to match?

File: pkg/controller/machine/machine_suite_test.go [hunk 8]
```diff
@@ -22,33 +22,33 @@ import (
...
 	shutdown := make(chan struct{})
 	si := sharedinformers.NewSharedInformers(config, shutdown)
-	controller := NewMachineController(config, si, fakeActuator{})
-	controller.Run(shutdown)
-
-	t.Run("machineControllerReconcile", func(t *testing.T) {
-		machineControllerReconcile(t, cs, controller)
-	})
-
+	actuator := &TestActuator{}
```
Comment (krousey) [inline pkg/controller/machine/machine_suite_test.go]: The creation of this actuator should be moved to the test method that uses it. Since it has internal state, it can't safely be reused between tests anyway.

File: pkg/controller/machine/machine_suite_test.go [hunk 9]
```diff
@@ -22,33 +22,33 @@ import (
...
+	controller := NewMachineController(config, si, actuator)
+	controller.RunAsync(shutdown)
+
+	// Create a cluster object.
+	cluster := testutil.GetVanillaCluster()
+	cluster.Name = "cluster-1"
+	if _, err := cs.ClusterV1alpha1().Clusters("default").Create(&cluster); err != nil {
```
Comment (krousey) [inline pkg/controller/machine/machine_suite_test.go]: Why is this here? Why don't the sub tests create the clusters if they need for their specific test?

File: pkg/controller/machine/machine_suite_test.go [hunk 10]
```diff
@@ -22,33 +22,33 @@ import (
...
+		t.Fatal(err)
+	}
+
+	// It's intentionally to run those test in sequence to simplify the test setup.
+	machineControllerReconcile(t, cs, controller)
```
Comment (krousey) [inline pkg/controller/machine/machine_suite_test.go]: I don't understand why you removed the calls to `t.Run`. They provide important separation of sub-tests both in testing framework (a Fatal in the first test can't stop the second from running), and in reporting.

File: pkg/controller/machine/testactuator.go [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+	a.Lock.Lock()
+	defer a.Lock.Unlock()
+	a.ExistsCallCount++
+	return a.ExistsValue, nil
+}
+
+func (a *TestActuator) DelayIfNeeded() {
+	if a.DelayInSeconds > 0 {
+		time.Sleep(time.Duration(a.DelayInSeconds) * time.Second)
```
Comment (krousey) [inline pkg/controller/machine/testactuator.go]: To echo @jessicaochen 's comment, do not rely on sleeps in test code.

File: pkg/controller/machine/machine_suite_test.go (hunk 9 above)
Comment (medinatiger) [reply pkg/controller/machine/machine_suite_test.go]: It's test case for machine resource, I think it's just common setup so that we can create machine.

File: pkg/controller/machine/machine_suite_test.go (hunk 10 above)
Comment (medinatiger) [reply pkg/controller/machine/machine_suite_test.go]: I'm having difficult to set up two test environment to run in parallel. I got following error. Do you have hint? 

 panic: the group "cluster.k8s.io" has already been announced.

goroutine 88 [running]:
sigs.k8s.io/cluster-api/vendor/k8s.io/apimachinery/pkg/apimachinery/announced.(*GroupMetaFactory).Announce(0xc420d8ef90, 0xc420335e90, 0xc420d8ef90)
	/usr/local/google/home/fmin/.gvm/pkgsets/go1.9.4/global/src/sigs.k8s.io/cluster-api/vendor/k8s.io/apimachinery/pkg/apimachinery/announced/group_factory.go:85 +0x6c

**Comment by medinatiger** [comment]
@krousey @jessicaochen rewrote the entire unittest according to some of the review comment. PTAL. Thanks!

File: pkg/controller/machine/controller_test.go [hunk 12]
```diff
@@ -94,3 +100,36 @@ func machineControllerReconcile(t *testing.T, cs *clientset.Clientset, controlle
...
+	for i := 0; i < 5; i++ {
+		instance := v1alpha1.Machine{}
+		instance.Name = "instance" + strconv.Itoa(i)
+		if _, err := client.Create(&instance); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	err := wait.Poll(time.Second, 10*time.Second, func() (bool, error) {
+		return (ta.CreateCallCount >= 2), nil
```
Comment (jessicaochen) [inline pkg/controller/machine/controller_test.go]: This should be 5 as all 5 of your instances should be processed.

File: pkg/controller/machine/machine_suite_test.go [hunk 13]
```diff
@@ -22,33 +22,31 @@ import (
...
 func TestMachine(t *testing.T) {
 	testenv := test.NewTestEnvironment()
 	config := testenv.Start(apis.GetAllApiBuilders(), openapi.GetOpenAPIDefinitions)
 	cs := clientset.NewForConfigOrDie(config)
 
 	shutdown := make(chan struct{})
 	si := sharedinformers.NewSharedInformers(config, shutdown)
-	controller := NewMachineController(config, si, fakeActuator{})
-	controller.Run(shutdown)
+	actuator := NewTestActuator()
```
[the remark this answers is not in the document]
Comment (medinatiger) [reply pkg/controller/machine/machine_suite_test.go]: Done

File: pkg/controller/machine/machine_suite_test.go [hunk 14]
```diff
@@ -20,35 +20,41 @@ import (
...
-	controller := NewMachineController(config, si, fakeActuator{})
-	controller.Run(shutdown)
-
+	// TODO: When cluster-api support per namepsace object, we need to make each subtest to run
+	// in different namespace. Everything lives inside default namespace so that the test needs to run
+	// sequentially, which is the default behavior for "go test".
 	t.Run("machineControllerReconcile", func(t *testing.T) {
+		controller, shutdown := getController(config)
 		machineControllerReconcile(t, cs, controller)
+		close(shutdown)
```
Comment (krousey) [inline pkg/controller/machine/machine_suite_test.go]: These should probably be deferred as a t.Fatal in the test would prevent this from running. You could also move controller creation inside of the test themselves to avoid having to to a type assertion on the actuator.

**Comment by krousey** [comment]
/hold
There's only one minor correction that I don't need to check. Remove the hold when fixed.
/lgtm

**Comment by krousey** [comment]
/lgtm
again
